src/runtime/aarch64_cpu_features.cpp

```diff
@@ -5,9 +5,140 @@ namespace Halide {
 namespace Runtime {
 namespace Internal {
 
+#if LINUX
+
+extern "C" unsigned long getauxval(unsigned long type);
+
+#define AT_HWCAP 16
+#define AT_HWCAP2 26
+
+// https://cs.android.com/android/platform/superproject/main/+/main:bionic/libc/kernel/uapi/asm-arm64/asm/hwcap.h
+// https://github.com/torvalds/linux/blob/master/arch/arm64/include/uapi/asm/hwcap.h
+#define HWCAP_ASIMDHP (1 << 10)
+#define HWCAP_ASIMDDP (1 << 20)
+#define HWCAP_SVE (1 << 22)
+#define HWCAP2_SVE2 (1 << 1)
+
+namespace {
+
+void set_platform_features(CpuFeatures &features) {
+    unsigned long hwcaps = getauxval(AT_HWCAP);
+    unsigned long hwcaps2 = getauxval(AT_HWCAP2);
+
+    if (hwcaps & HWCAP_ASIMDDP) {
+        features.set_available(halide_target_feature_arm_dot_prod);
+    }
+
+    if (hwcaps & HWCAP_ASIMDHP) {
+        features.set_available(halide_target_feature_arm_fp16);
+    }
+
+    if (hwcaps & HWCAP_SVE) {
+        features.set_available(halide_target_feature_sve);
+    }
+
+    if (hwcaps2 & HWCAP2_SVE2) {
+        features.set_available(halide_target_feature_sve2);
+    }
+}
+
+}  // namespace
+
+#elif OSX
+
+extern "C" int sysctlbyname(const char *name, void *oldp, size_t *oldlenp, void *newp, size_t newlen);
+
+namespace {
+
+bool sysctl_is_set(const char *name) {
+    int enabled = 0;
+    size_t enabled_len = sizeof(enabled);
+    return sysctlbyname(name, &enabled, &enabled_len, nullptr, 0) == 0 && enabled;
+}
+
+void set_platform_features(CpuFeatures &features) {
+    if (sysctl_is_set("hw.optional.arm.FEAT_DotProd")) {
+        features.set_available(halide_target_feature_arm_dot_prod);
+    }
+
+    if (sysctl_is_set("hw.optional.arm.FEAT_FP16")) {
+        features.set_available(halide_target_feature_arm_fp16);
+    }
+
+    // Apple M3 sets FEAT_SME to 0 and does not define FEAT_SVE
```

Comment:
This too

Reply:
Removed -- yeah, I thought that ARMv9 was going to require "SME implies SVE"